### ISSUE TITLE
`terraform test`: Additional erroneous error message when referencing invalid attributes

### ISSUE THREAD

**Issue by liamcervante** [hashicorp/terraform#35265]
### Terraform Version

```shell
v1.8.3
```

### Terraform Configuration Files

```terraform
# main.tf

variable "number" {
    type = number
}

data "external" "resource" {
    count = var.number

    program = [ "bash", "${path.module}/out.sh" ]
}

output "out" {
    value = data.external.resource
}
```

```bash
#!/bin/bash
# out.sh

echo "{\"hello\":\"world\"}"
```

```terraform
# main.tftest.hcl

run "test" {
    variables {
        number = 1
    }

    assert {
        condition = data.external.resource[0].hello == "world"
        error_message = "wrong"
    }
}
```

### Debug Output

N/A

### Expected Behavior

```
main.tftest.hcl... in progress
  run "test"... fail
╷
│ Error: Unsupported attribute
│ 
│   on main.tftest.hcl line 8, in run "test":
│    8:         condition = data.external.resource[0].hello == "world"
│ 
│ This object has no argument, nested block, or exported attribute named "hello".
╵
main.tftest.hcl... tearing down
main.tftest.hcl... fail

Failure! 0 passed, 1 failed.
```

### Actual Behavior

```
main.tftest.hcl... in progress
  run "test"... fail
╷
│ Error: Unknown variable
│ 
│   on main.tftest.hcl line 8, in run "test":
│    8:         condition = data.external.resource[0].hello == "world"
│ 
│ There is no variable named "data".
╵
╷
│ Error: Unsupported attribute
│ 
│   on main.tftest.hcl line 8, in run "test":
│    8:         condition = data.external.resource[0].hello == "world"
│ 
│ This object has no argument, nested block, or exported attribute named "hello".
╵
main.tftest.hcl... tearing down
main.tftest.hcl... fail

Failure! 0 passed, 1 failed.
```

### Steps to Reproduce

1. `terraform init`
2. `terraform test`

### References

https://discuss.hashicorp.com/t/tf-test-1-8-cannot-test-custom-data-source-output/67461

**Comment by liamcervante** [comment]
We are constructing a HCL evaluation context for the test command [here](https://github.com/hashicorp/terraform/blob/main/internal/moduletest/eval_context.go#L95). In this case, the actual construction of the context does some validation of its own which results in it returning the expected diagnostic along with an empty context. The error then comes as we are not exiting the evaluation early when the context isn't computed properly. Currently, we are just adding the diagnostics and then attempting to evaluate the condition anyway. This results in the confusing message we get as the context does not contain any variables.

The fix would be to skip the evaluation of the rule if we cannot build the context properly:

```golang
		hclCtx, moreDiags := scope.EvalContext(refs)
		ruleDiags = ruleDiags.Append(moreDiags)
                if moreDiags.HasErrors() {
                    // If we couldn't create the context properly then we can't
                    // evaluate the condition.
                    continue
                }
```

**Comment by burnerlee** [comment]
Hey can i take this issue up?

**Comment by liamcervante** [comment]
Hi @burnerlee, I'll happily review a PR for this if you put one together.

**Comment by burnerlee** [comment]
thanks

**Comment by burnerlee** [comment]
@liamcervante opened a PR for the issue

**Comment by liamcervante** [comment]
This is fixed and merged, will be released in v1.9.1. Thanks @burnerlee
